packages/runtime/container-runtime/src/summarizer.ts

```diff
@@ -252,13 +261,34 @@ export class Summarizer extends EventEmitter implements ISummarizer {
         }
     }
 
-    public readonly summarizeOnDemand: ISummarizer["summarizeOnDemand"] = (...args) => {
-        if (this._disposed || this.runningSummarizer === undefined || this.runningSummarizer.disposed) {
-            throw Error("Summarizer is not running or already disposed.");
+    public readonly summarizeOnDemand: ISummarizer["summarizeOnDemand"] = async (...args) => {
+        try {
+            if (this._disposed) {
+                throw Error("Summarizer is already disposed.");
+            }
+            if (this.runningSummarizer === undefined || this.runningSummarizer.disposed) {
+                this.runningSummarizer = undefined;
+                const runCoordinator: ICancellableSummarizerController =
+                    await this.runCoordinatorCreateFn(this.runtime);
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                await this.start(this.runtime.clientId!, runCoordinator, { disableHeuristics: true });
+            }
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            return this.runningSummarizer!.summarizeOnDemand(...args);
+        }
+        catch (error) {
+            throw SummarizingWarning.wrap(error, "summarizerRun", false /* logged */, this.logger);
         }
-        return this.runningSummarizer.summarizeOnDemand(...args);
     };
 
+    public async stopOnDemand() {
```

Comment:
It would be better (from API shape perspective) if summarizeOnDemand() API was returning an object that would have stop() method or something of that nature. Having it like that seems really weird, as anybody can call it at any time and we need to clearly articulate and document behavior (and I think it will cancel not just on-demand summary, but any running summary).

Reply:
Comments in the code (and possibly a different name) would probably help. The idea was not to stop a particular summary (though that would happen also) but to shut down the on-demand summarizer when it's no longer needed. Basically, do the reverse of what start() does.

Reply:
No, this should be done another way. It should be possible to use the existing summarizer.stop() and have the resolution of the promise lead to the tearing down of the summarizer, the way it does in the SummaryManager flow. Working on that.